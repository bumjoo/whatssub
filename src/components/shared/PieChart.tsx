import React, { Component } from 'react';
import {
  View,
  Text,
  ViewStyle,
} from 'react-native';

import styled from 'styled-components/native';
import { PieChart } from 'react-native-svg-charts';
import { ColorProperty } from 'csstype';
import { colors } from '../../theme';
import { getString } from '../../../STRINGS';

export interface IProduct {
  name: string;
  category: string;
  price: number;
  currency: string;
  date: Date;
  image: string;
  currentMonthPaymentDate: Date;
};

const Container = styled.View`
  background-color: white;
  align-items: center;
  justify-content: center;
`;
const StyledViewInfo = styled.View`
  background-color: transparent;
  flex-direction: column;
`;
const StyledText = styled.Text`
  font-size: 14;
  color: ${({ theme }) => theme.btnPrimaryFont};
`;
const StyledTextMonth = styled.Text`
  font-size: 14;
  color: ${({ theme }) => theme.fontDark};
`;
const StyledTextInfoCategory = styled.Text`
  font-size: 14;
  color: 'rgb(62,126,255)';
`;
const StyledTextInfoValue = styled.Text`
  font-size: 14;
  color: ${colors.gray};
`;

type IProps = {
  style?: ViewStyle;
  data: IProduct[];
  defaultShowSlice?: number;
  currentMonth: string;
}

const getCategoryPrice = (data: IProduct[]) => {
  const categoryPrice = {};
  data && data.forEach((item:IProduct) => {
    if (!categoryPrice[item.category]) categoryPrice[item.category] = 0;
    categoryPrice[item.category] = categoryPrice[item.category] + item.price;
  });
  return categoryPrice;
};
const initSelectedSlice = { key: '', value: 0 };

export default function SharedPieChart(props: IProps) {
  const [selectedSlice, setSelectedSlice] = React.useState(initSelectedSlice);
  // Total Amount by Category
  const categoryPrice = React.useMemo(() => {
    return getCategoryPrice(props.data);
  }, [props.data]);
  // Total Amount
  const totalPrice = React.useMemo(() => {
    const total = props.data && props.data.reduce((acc, item) => {
      return acc + item.price;
    }, 0);
    return total;
  }, [props.data]);
  const currency = React.useMemo(() => {
    return (props.data && props.data[0].currency) || '원';
  }, [props.data]);

  const pieData = categoryPrice && Object.keys(categoryPrice).map((category, index) => {
    const color = (selectedSlice.key === category) ? 'rgb(62,126,255)'
      : index % 3 === 0 ? '#666666' : index % 3 === 1 ? '#999999' : '#cccccc';
    return {
      key: category,
      value: categoryPrice[category],
      svg: { fill: color },
      arc: {
        innerRadius: (selectedSlice.key === category ? '75%' : '80%'),
        outerRadius: (selectedSlice.key === category ? '100%' : '90%'),
        padAngle: (selectedSlice.key === category ? 0 : 0.02),
      },
      onPress: () => {
        setSelectedSlice({
          key: category,
          value: categoryPrice[category],
        });
      },
    };
  });

  React.useEffect(() => {
    if (props.defaultShowSlice >= 0) {
      const num = props.defaultShowSlice > (pieData.length - 1) ? 0 : props.defaultShowSlice;
      (pieData.length > 0) && setSelectedSlice({
        key: pieData[num].key,
        value: pieData[num].value,
      });
    }
  }, [props.data]);
  if (pieData && pieData.length > 0) {
    return (
      <Container style={props.style}>
        {/** Data View for Selected Slice */}
        <StyledViewInfo style={{ position: 'absolute', left: '75%', top: 0, width: 100, height: 50 }}>
          <StyledTextInfoCategory>
            {`${selectedSlice.key}`}
          </StyledTextInfoCategory>
          <StyledTextInfoValue>
            {`${selectedSlice.value}${currency}`}
          </StyledTextInfoValue>
        </StyledViewInfo>
        {/** Donut chart */}
        <PieChart
          style={{ width: 214, height: 214 }}
          data={pieData}
          spacing={0}
          // outerRadius={'100%'}
          // innerRadius={'90%'}
          // valueAccessor={({ item }) => item.amount}
        >
          {/* <Labels slices={null} height={null} width={null}/> */}
        </PieChart>
        {/** centered elements : month, total amount to be paid,  */}
        <View style={{ position: 'absolute', justifyContent: 'center', alignItems: 'center' }}>
          <StyledTextMonth style={{}}>
            {`${props.currentMonth || '2016.00'}`}
          </StyledTextMonth>
          <StyledText style={{ fontSize: 14, color: colors.gray }}>
            {getString('AMOUNT_TO_BE_PAID')}
          </StyledText>
          <StyledText style={{ fontSize: 24, color: colors.darkGray }}>
            {`${totalPrice}${currency}`}
          </StyledText>
          {/* <StyledText style={{ fontSize: 20 }}>{`${selectedSlice.key} \n ${selectedSlice.value}`}</StyledText> */}
        </View>
      </Container>
    );
  } else {
    return null;
  }
}
