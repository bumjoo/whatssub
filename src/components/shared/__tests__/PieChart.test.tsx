import 'react-native';
import * as React from 'react';
import PieChart from '../PieChart';

import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';
import { render, fireEvent, act, RenderResult } from '@testing-library/react-native';

import { createTheme, ThemeType } from '../../../theme';

const createTestProps = (obj: object) => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...obj,
});
const props: any = createTestProps({});
const component = (
  <ThemeProvider theme={createTheme(ThemeType.LIGHT)}>
    <PieChart {...props} />
  </ThemeProvider>
);

describe('[PieChart]', () => {
  it('renders without crashing', () => {
    const rendered = renderer.create(component).toJSON();
    expect(rendered).toMatchSnapshot();
    // expect(rendered).toBeTruthy();
  });
});
