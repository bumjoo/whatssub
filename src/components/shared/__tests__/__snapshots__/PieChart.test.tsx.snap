// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[PieChart] renders without crashing 1`] = `null`;
